Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewRowCollection.cs

```diff
@@ -0,0 +1,614 @@
+            Contract.Requires(this.DataGridView.DataSource == null);
+            Contract.Requires(this.DataGridView.RowTemplate.Cells.Count <= this.DataGridView.Columns.Count);
+        }
+        
+        /// <summary>
+        /// Inserts a row into the collection at the specified position, based on the row at specified position.
+        /// </summary>
+        /// <param name="indexSource">The index of the row on which to base the new row.</param><param name="indexDestination">The position at which to insert the row.</param><exception cref="T:System.ArgumentOutOfRangeException"><paramref name="indexSource"/> is less than zero or greater than the number of rows in the collection minus one.-or-<paramref name="indexDestination"/> is less than zero or greater than the number of rows in the collection.</exception><exception cref="T:System.InvalidOperationException">The associated <see cref="T:System.Windows.Forms.DataGridView"/> control is performing one of the following actions that temporarily prevents new rows from being added:Selecting all cells in the control.Clearing the selection.-or-This method is being called from a handler for one of the following <see cref="T:System.Windows.Forms.DataGridView"/> events:<see cref="E:System.Windows.Forms.DataGridView.CellEnter"/><see cref="E:System.Windows.Forms.DataGridView.CellLeave"/><see cref="E:System.Windows.Forms.DataGridView.CellValidating"/><see cref="E:System.Windows.Forms.DataGridView.CellValidated"/><see cref="E:System.Windows.Forms.DataGridView.RowEnter"/><see cref="E:System.Windows.Forms.DataGridView.RowLeave"/><see cref="E:System.Windows.Forms.DataGridView.RowValidated"/><see cref="E:System.Windows.Forms.DataGridView.RowValidating"/>-or-<paramref name="indexDestination"/> is equal to the number of rows in the collection and <see cref="P:System.Windows.Forms.DataGridView.AllowUserToAddRows"/> is true. -or-This operation would insert a frozen row after unfrozen rows or an unfrozen row before frozen rows.</exception><PermissionSet><IPermission class="System.Security.Permissions.EnvironmentPermission, mscorlib, Version=2.0.3600.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" version="1" Unrestricted="true"/><IPermission class="System.Security.Permissions.FileIOPermission, mscorlib, Version=2.0.3600.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" version="1" Unrestricted="true"/><IPermission class="System.Security.Permissions.SecurityPermission, mscorlib, Version=2.0.3600.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" version="1" Flags="UnmanagedCode, ControlEvidence"/><IPermission class="System.Diagnostics.PerformanceCounterPermission, System, Version=2.0.3600.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" version="1" Unrestricted="true"/></PermissionSet>
+        public virtual void InsertCopy(int indexSource, int indexDestination)
+        {
+            this.InsertCopies(indexSource, indexDestination, 1);
```

Comment:
Why no preconditions? I don't think that preconditions could be 'inherited' by calling another method.